### PR TITLE
Add WebSocket channel auto-resubscription

### PR DESCRIPTION
**Pull request by hjoelr**
@bennycode Here is a rough draft of #421. I've not written tests or even run the code yet. I just didn't want to go through full validation of it without getting your input on the architecture I've gone with. Could you take a look at what I have so far and give me any direction on changes you'd like to see?

In essence I've created a hashmap by channel name and keep the WebSocketChannel product IDs updated as subscribe and unsubscribe functions are called. If disconnect is called or the connection drops, the default behavior is to remember the previous subscriptions and re-subscribe to them when the connection comes back up. I added an optional parameter to the disconnect function to intentionally forget subscriptions if desired.

### CODE REVIEW

**Comment by hjoelr** [comment]
@bennycode I added tests and updated the code to fix some bugs I had in my original version. If you would like changes or improvements in the API of this PR, I'm happy to make the changes and re-submit.

Also, you'll notice that the code coverage is failing because of some branches missing. However, I'm not sure how to fix that because I don't understand why those lines are failing branch coverage. Could you give me some direction on that?

File: src/client/WebSocketClient.ts [hunk 1]
```diff
@@ -464,7 +478,12 @@ export class WebSocketClient extends EventEmitter {
     return this.socket;
   }
 
-  disconnect(reason: string = 'Unknown reason'): void {
+  disconnect(reason: string = 'Unknown reason', forgetSubscriptions: boolean = false): void {
```
Comment (hjoelr) [inline src/client/WebSocketClient.ts:481]: I'm not too thrilled about this function signature because giving just the `forgetSubcriptions` without a reason requires something like this:

```TypeScript
ws.disconnect(undefined, true);

// OR

ws.disconnect('', true);
```

Maybe another option would be to convert the parameters to an options object like so:

```TypeScript
disconnect(options?: {
  reason?: boolean,
  forgetSubscriptions?: boolean,
}): void
```

We could accomplish that without changing the API with some checks if the param is a string or an object.

File: src/client/WebSocketClient.ts [hunk 2]
```diff
@@ -364,11 +364,19 @@ export class WebSocketClient extends EventEmitter {
   private pingTime: number;
   private readonly pongTime: number;
 
+  private _subscriptions: {
+    [channel_name: string]: Required<WebSocketChannel>;
```
Comment (hjoelr) [inline src/client/WebSocketClient.ts:368]: I don't know of a way to limit an index signature object to just the values of an enum (`WebSocketChannelName`). Maybe you know how to wrangle TypeScript to do that?

I used index signature as my master store because I believe its must more performant than continually looping through an array to find items.

File: src/client/WebSocketClient.ts [hunk 3]
```diff
@@ -364,11 +364,19 @@ export class WebSocketClient extends EventEmitter {
...
+  };
+
+  private _subscriptions_cache: WebSocketChannel[];
```
Comment (hjoelr) [inline src/client/WebSocketClient.ts:371]: This exists mainly for performance optimization so that every time ws.subscriptions is called we don't have to regenerate this array. It's hard to know if ws.subscriptions is going to be called more in the wild or ws.subscribe()/unsubscribe().

File: src/client/WebSocketClient.ts [hunk 4]
```diff
@@ -502,19 +521,117 @@ export class WebSocketClient extends EventEmitter {
...
+      }
+      return -1; // assume a's name is less than b's name
+    });
+  }
+
+  /**
+   * Returns all the channels and products that have been subscribed.
+   */
+  get subscriptions(): WebSocketChannel[] {
+    return this._subscriptions_cache;
```
Comment (hjoelr) [inline src/client/WebSocketClient.ts:632]: This cache array should be kept up-to-date as changes happen, so we should just need to return it.

File: src/client/WebSocketClient.test.ts [hunk 5]
```diff
@@ -219,6 +219,96 @@ describe('WebSocketClient', () => {
...
+
+          if (request.type === WebSocketRequestType.UNSUBSCRIBE) {
+            // Send unsubscribe confirmation
+            server.clients.forEach(client => client.send(JSON.stringify(emptySubscriptions)));
+          }
+        });
+      });
+    }
+
+    function mockWebSocketClientSubscription(
```
Comment (hjoelr) [inline src/client/WebSocketClient.test.ts:252]: I wrote this function so that it could replace the `mockWebSocketResponse()` function. However, I did not replace that function in this PR as that's out of the scope of this PR. The `mockWebSocketResponse()` could not work for my needs here because it makes some assumptions that wouldn't allow me to properly test this functionality.

**Comment by hjoelr** [comment]
After looking further into the code coverage issues, it appears that these are edge cases that nyc can't identify correctly. In one case where I use `hasOwnProperty`, that's a [known scenario](https://github.com/gotwarlost/istanbul/blob/master/ignoring-code-for-coverage.md) where it's not accurate. In the other, the product_ids property should never be undefined because we wouldn't have gotten to that code if it was. However, we still must use the `?.` operator which I think is what it doesn't like.

On a different note, I was wondering if we should add either a constructor parameter to designate whether or not we want WebSocketClient to automatically resubscribe...

```TypeScript
constructor(baseURL: string, private readonly signRequest: (setup: RequestSetup) => Promise<SignedRequest>, autoResubscribe: boolean = true)
```

...or maybe add it as an option in the connect function:

```TypeScript
connect(reconnectOptions?: Options & { autoResubscribe: boolean }): ReconnectingWebSocket
```

...or, have it default to resubscribe but add a function called `disableResubscribe()` that can be called anytime during the `WebSocketClient`'s lifecycle.

Each has merit, but I probably lean toward the `disableResubscribe()` option. Anyway, interested in your thoughts. Sorry for being so prolific!

**Comment by bennycode** [comment]
Hi @hjoelr. No need to be sorry! I am very happy that you put so much thought in this library. Can you exchange `?.` with `!.`? I am infavor of adding a param to `connect` to decide if we should re-subscribe or not. Thus users do not have to commit themselves from the beginning and can change their re-subscription easily on a second connect.

**Comment by hjoelr** [comment]
@bennycode Okay, I made those tweaks as you suggested. I'm running up against another code coverage issue that I don't know how to solve. Any ideas why it's saying line 405 isn't covering a branch and how I can fix it?

Also, I was trying to write a test to make sure it doesn't **actually** resubscribe upon reconnect. However, I couldn't figure out how to do it since I'm basically needing to test something like "make sure the subscription callback doesn't run more than once". I'm not sure how to do that. Any ideas?

**Comment by bennycode** [comment]
Is this line 405 that you want to test?

https://github.com/bennycode/coinbase-pro-node/blob/48d4dbc04754c5ed28acb95feb8fe3506703f99d/src/client/WebSocketClient.ts#L405

If yes, then the coverage tool wants you to test the positive case and the negative case for that if-condition. When you test only one of the two, then it will complain with "isn't covering a branch".

Regarding the subscription callback... Can you work with something like `once`? https://nodejs.org/api/events.html#emitteronceeventname-listener

**Comment by hjoelr** [comment]
@bennycode This PR is ready for final review. Please let me know if there's anything else you think I should do on it.

File: src/client/WebSocketClient.ts [hunk 6]
```diff
@@ -494,7 +499,26 @@ export class WebSocketClient extends EventEmitter {
...
+    /*
+     * To prevent breaking the legacy API, we allow options to be a string to represent
+     * the reason for the disconnect.
+     */
+    if (typeof options === 'string') {
+      reason = options;
+    } else if (options) {
+      if (undefined !== options.forgetSubscriptions) {
+        forgetSubscriptions = options.forgetSubscriptions;
+      }
```
Comment (hjoelr) [inline src/client/WebSocketClient.ts:515]: I wish I could change this to just this...

```TypeScript
forgetSubscriptions = options?.forgetSubscriptions ?? false;
```

However, I can't figure out how to do coverage on that.

File: src/client/WebSocketClient.ts (hunk 1 above)
Comment (hjoelr) [reply src/client/WebSocketClient.ts]: I'm sure you saw it, but I pushed a commit that changes the parameter to `options` while still allowing it to be a string indicating the disconnect reason for backward compatibility.

**Comment by bennycode** [comment]
Can you rebase it with the "main" branch? I will try to review it then next week (I am off for the weekend).

**Comment by hjoelr** [comment]
@bennycode I've had to set this project aside for a while, which is why I've been a little silent here. Also, I've been contemplating if this change is actually beneficial. I was finding in my case that I would have to keep a separate store of requested subscriptions anyway as I want to have a way to identify which part of my program is requesting to subscribe/unsubscribe and only **actually** unsubscribe from a product if every part in my program that requested to subscribe also requested to unsubscribe. This means I need to store some token to identify which part of the program requested the subscribe. The PR as it stands doesn't currently support something like this.

This PR could be reworked to include such functionality, but then it makes me wonder if that's really outside the scope of this library. Anyway, I'd be interested if you have any thoughts on that.

**Comment by bennycode** [comment]
Hi @[hjoelr](https://github.com/hjoelr), unfortunately I didn't check your PR as expected. Is it still relevant? Can we move portions of it into new PRs to merge it bit-by-bit without breaking any API contract?

**Comment by bennycode** [comment]
Closing this because of inactivity. 😶‍🌫️

@hjoelr just tell me if you plan to get back to it and I will take the time to support you. 💪
